pkg/coordinator/redemptions.go

```diff
@@ -0,0 +1,328 @@
+	if fee <= 0 {
+		panic("fee estimation not implemented yet")
+	}
+
+	logger.Infof("redemption transaction fee: [%d]", fee)
+
+	logger.Infof("preparing a redemption proposal...")
+
+	proposal := &tbtc.RedemptionProposal{
+		WalletPublicKeyHash:    walletPublicKeyHash,
+		RedeemersOutputScripts: redeemersOutputScripts,
+		RedemptionTxFee:        big.NewInt(fee),
+	}
+
+	logger.Infof("validating the redemption proposal...")
+
+	if _, err := tbtc.ValidateRedemptionProposal(
+		logger,
+		proposal,
+		chain,
+	); err != nil {
+		return fmt.Errorf("failed to verify redemption proposal: %v", err)
+	}
+
+	if !dryRun {
+		logger.Infof("submitting the deposit sweep proposal...")
+		if err := chain.SubmitRedemptionProposalWithReimbursement(proposal); err != nil {
+			return fmt.Errorf("failed to submit deposit sweep proposal: %v", err)
+		}
+	} else {
+		logger.Infof("skipping transaction submission in dry-run mode")
+	}
+
+	return nil
+}
+
+func getPendingRedemptions(
+	chain Chain,
+	walletPublicKeyHash [20]byte,
+	maxNumberOfRedemptions int,
+	redemptionRequestTimeout uint32,
+	redemptionRequestMinAge uint32,
+) ([]*redemptionEntry, error) {
+	logger.Infof("reading revealed deposits from chain...")
+
+	filter := &tbtc.RedemptionRequestedEventFilter{}
+	if walletPublicKeyHash != [20]byte{} {
+		filter.WalletPublicKeyHash = [][20]byte{walletPublicKeyHash}
+	}
+
+	// Only redemption requests in range:
+	// [now - redemptionRequestTimeout, now - redemptionRequestMinAge]
+	// should be taken into consideration.
+	redemptionRequestsRangeStartTimestamp := time.Now().Add(
+		time.Duration(-redemptionRequestTimeout) * time.Second,
+	)
+	redemptionRequestsRangeEndTimestamp := time.Now().Add(
+		time.Duration(-redemptionRequestMinAge) * time.Second,
+	)
+
+	// TODO: We should consider narrowing the block range in filter the fetch
+	//       only events that are newer than `now - redemptionRequestTimeout`.
+
+	redemptionsRequested, err := chain.PastRedemptionRequestedEvents(filter)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"failed to get past redemption requested events: [%w]",
+			err,
+		)
+	}
+	logger.Infof("found %d RedemptionRequested events", len(redemptionsRequested))
+
+	// Take the oldest first.
+	sort.SliceStable(redemptionsRequested, func(i, j int) bool {
+		return redemptionsRequested[i].BlockNumber < redemptionsRequested[j].BlockNumber
+	})
+
+	logger.Infof("checking pending redemptions details...")
+
+	pendingRedemptions := make([]*redemptionEntry, 0)
+
+redemptionRequestedLoop:
+	for i, event := range redemptionsRequested {
+		logger.Debugf("getting pending redemption details %d/%d", i+1, len(redemptionsRequested))
+
+		// Check if there is still a pending redemption for the given redemption
+		// requested event.
+		pendingRedemption, err := chain.GetPendingRedemptionRequest(
+			event.WalletPublicKeyHash,
+			event.RedeemerOutputScript,
+		)
+		if err != nil {
+			switch {
+			case errors.Is(err, tbtc.ErrPendingRedemptionRequestNotFound):
```

Comment:
This is where we could benefit from the additional `bool` returned from the `GetPendingRedemptionRequest` 😉

Reply:
On the first look - yes. On the other hand, in the most common case where we need simple info whether the request exists or not we would be forced to check the additional bool flag. Additionally, this method starts with `Get*` which implies that we either get the request if it exists or fail otherwise. I think the bool flag would be helpful for some `Find*` method where the non-existing result is a typical case. That said, I'm not sure if it's worth complicating the API of that method.